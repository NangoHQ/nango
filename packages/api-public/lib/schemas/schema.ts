import z from 'zod';

export const apiSchemaRegistry = z.registry<{ id: string }>();
