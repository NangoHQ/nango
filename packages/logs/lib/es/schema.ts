import type { estypes } from '@elastic/elasticsearch';
import type { MessageRow, OperationRow } from '@nangohq/types';
import { envs } from '../env.js';

const props: Record<keyof MessageRow | keyof OperationRow, estypes.MappingProperty> = {
    id: { type: 'keyword' },

    parentId: { type: 'keyword' },

    accountId: { type: 'keyword' },
    accountName: { type: 'keyword' },

    environmentId: { type: 'keyword' },
    environmentName: { type: 'keyword' },

    integrationId: { type: 'keyword' },
    integrationName: {
        type: 'text',
        analyzer: 'standard',
        search_analyzer: 'standard',
        fields: {
            keyword: {
                type: 'keyword'
            }
        }
    },
    providerName: { type: 'keyword' },

    connectionId: { type: 'keyword' },
    connectionName: {
        type: 'text',
        analyzer: 'standard',
        search_analyzer: 'standard',
        fields: {
            keyword: {
                type: 'keyword'
            }
        }
    },
    endUserId: { type: 'keyword' },
    endUserName: { type: 'keyword' },

    syncConfigId: { type: 'keyword' },
    syncConfigName: {
        type: 'text',
        analyzer: 'standard',
        search_analyzer: 'standard',
        fields: {
            keyword: {
                type: 'keyword'
            }
        }
    },

    jobId: { type: 'keyword' },

    userId: { type: 'keyword' },

    operation: {
        properties: {
            type: { type: 'keyword' },
            action: { type: 'keyword' }
        }
    },

    type: { type: 'keyword' },
    level: { type: 'keyword' },
    state: { type: 'keyword' },
    context: { type: 'keyword' },

    source: { type: 'keyword' },

    message: { type: 'text', analyzer: 'standard', search_analyzer: 'standard' },
    meta: { type: 'object', enabled: false },
    error: { type: 'object', enabled: false },

    request: {
        properties: {
            url: { type: 'keyword' },
            method: { type: 'keyword' },
            headers: { type: 'object', enabled: false }
        }
    },
    response: {
        properties: {
            code: { type: 'integer' },
            headers: { type: 'object', enabled: false }
        }
    },
    retry: {
        properties: {
            max: { type: 'integer' },
            attempt: { type: 'integer' },
            waited: { type: 'integer' }
        }
    },

    createdAt: { type: 'date' },
    updatedAt: { type: 'date' },
    startedAt: { type: 'date' },
    expiresAt: { type: 'date' },
    endedAt: { type: 'date' },
    durationMs: { type: 'integer' }
};

export function getDailyIndexPipeline(name: string): estypes.IngestPutPipelineRequest {
    return {
        id: `daily.${name}`,
        description: 'Daily index',
        processors: [
            {
                date_index_name: {
                    field: 'createdAt',
                    index_name_prefix: `${name}.`,
                    date_rounding: 'd',
                    date_formats: ["yyyy-MM-dd'T'HH:mm:ss.SSSXX"]
                }
            }
        ]
    };
}

export function policyRetention(): estypes.IlmPutLifecycleRequest {
    return {
        name: 'policy_retention',
        policy: {
            phases: {
                delete: { min_age: '15d', actions: { delete: {} } }
            }
        }
    };
}

export const indexMessages: estypes.IndicesCreateRequest = {
    index: `20240528_${envs.NANGO_LOGS_ES_INDEX ?? 'messages'}`,
    settings: {
        lifecycle: { name: 'policy_retention' },
        analysis: {
            analyzer: {
                default: {
                    type: 'standard'
                },
                default_search: {
                    type: 'standard'
                }
            }
        },
        index: {
            'sort.field': ['createdAt', 'id'],
            'sort.order': ['desc', 'desc']
        },
        // They are recommending 1 shard per 20gb-40gb
        number_of_shards: envs.NANGO_LOGS_ES_SHARD_PER_DAY
    },
    mappings: {
        _source: { enabled: true },
        dynamic: false,
        properties: props
    }
};
