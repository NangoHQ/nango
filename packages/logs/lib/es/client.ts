import { Client } from '@elastic/elasticsearch';

import { envs } from '../env.js';

export const client = new Client({
    nodes: envs.NANGO_LOGS_ES_URL || 'http://localhost:0',
    requestTimeout: 5000,
    maxRetries: 1,
    auth: {
        username: envs.NANGO_LOGS_ES_USER!, // ggignore
        password: envs.NANGO_LOGS_ES_PWD! // ggignore
    }
});
