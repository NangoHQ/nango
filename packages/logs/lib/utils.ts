import { errorToObject, getLogger } from '@nangohq/utils';
import { createKVStore } from '@nangohq/kvstore';
import type { KVStore } from '@nangohq/kvstore';
import type { MessageRow } from '@nangohq/types';

export const logger = getLogger('logs');

export const isCli = process.argv.find((value) => value.includes('/bin/nango') || value.includes('cli/dist/index'));

let kvstorePromise: Promise<KVStore> | undefined;
export async function getKVStore(): Promise<KVStore> {
    if (!kvstorePromise) {
        kvstorePromise = createKVStore();
        return await kvstorePromise;
    }

    return await kvstorePromise;
}

export async function destroy() {
    logger.info('Destroying logs...');
    if (kvstorePromise) {
        await (await kvstorePromise)?.destroy();
    }
    kvstorePromise = undefined;
}

export const logLevelToLogger = {
    info: 'info',
    debug: 'debug',
    error: 'error',
    warn: 'warning',
    http: 'info',
    verbose: 'debug',
    silly: 'debug'
} as const;

export function errorToDocument(error?: unknown): MessageRow['error'] {
    if (!error) {
        return null;
    }

    const err = { name: 'Unknown Error', message: 'unknown error', ...errorToObject(error) };
    return {
        name: error instanceof Error ? error.constructor.name : err.name,
        message: err.message,
        type: 'type' in err ? (err.type as string) : null,
        payload: 'payload' in err ? err.payload : null
    };
}
