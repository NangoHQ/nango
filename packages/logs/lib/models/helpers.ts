import { nanoid } from '@nangohq/utils';
import type { ConcatOperationList, MessageRow, MessageRowInsert, OperationRow, OperationRowInsert } from '@nangohq/types';
import { z } from 'zod';
import type { estypes } from '@elastic/elasticsearch';
import { defaultOperationExpiration } from '../env.js';
import type { LogContext } from '../client.js';

export const operationIdRegex = z.string().regex(/([0-9]|[a-zA-Z0-9]{20})/);

export interface FormatMessageData {
    account?: { id: number; name: string };
    user?: { id: number } | undefined;
    environment?: { id: number; name: string } | undefined;
    connection?: { id: number; name: string } | undefined;
    integration?: { id: number; name: string; provider: string } | undefined;
    syncConfig?: { id: number; name: string } | undefined; // TODO: rename to script or something similar because it also apply to actions and on-events scripts
    meta?: MessageRow['meta'];
}

export function getFormattedOperation(
    data: OperationRowInsert,
    { account, user, environment, integration, connection, syncConfig, meta }: FormatMessageData = {}
): OperationRow {
    return {
        ...getFormattedMessage(data as unknown as MessageRow),
        message: operationTypeToMessage[`${data.operation.type}:${data.operation.action}` as ConcatOperationList],
        id: data.id || nanoid(),
        operation: data.operation,

        accountId: account?.id ?? data.accountId ?? -1,
        accountName: account?.name || data.accountName || '',

        environmentId: environment?.id ?? data.environmentId ?? null,
        environmentName: environment?.name || data.environmentName || null,

        integrationId: integration?.id ?? data.integrationId ?? null,
        integrationName: integration?.name || data.integrationName || null,
        providerName: integration?.provider || data.providerName || null,

        connectionId: connection?.id ?? data.connectionId ?? null,
        connectionName: connection?.name || data.connectionName || null,

        syncConfigId: syncConfig?.id || data.syncConfigId || null,
        syncConfigName: syncConfig?.name || data.syncConfigName || null,

        jobId: data.jobId || null,
        meta: meta || data.meta || null,

        userId: user?.id || data.userId || null,
        parentId: null
    };
}
export function getFormattedMessage(data: Partial<MessageRow>, { meta }: FormatMessageData = {}): MessageRow {
    const now = new Date();
    return {
        id: data.id || nanoid(), // This ID is for debugging purpose, not for insertion

        source: data.source || 'internal',
        level: data.level || 'info',
        operation: data.operation || null,
        type: data.type || 'log',
        message: data.message || '',
        title: data.title || null,
        code: data.code || null,
        state: data.state || 'waiting',

        accountId: null,
        accountName: null,

        environmentId: null,
        environmentName: null,

        integrationId: null,
        integrationName: null,
        providerName: null,

        connectionId: null,
        connectionName: null,

        syncConfigId: null,
        syncConfigName: null,

        jobId: data.jobId || null,

        userId: null,
        parentId: data.parentId || null,

        error: data.error || null,
        request: data.request || null,
        response: data.response || null,
        meta: meta || data.meta || null,

        createdAt: data.createdAt || now.toISOString(),
        updatedAt: data.updatedAt || now.toISOString(),
        startedAt: data.startedAt || null,
        endedAt: data.endedAt || null,
        expiresAt: data.operation ? data.expiresAt || defaultOperationExpiration.sync() : null
    };
}

// TODO: remove once not used by persist anymore
export const oldLevelToNewLevel = {
    debug: 'debug',
    info: 'info',
    warn: 'warn',
    error: 'error',
    verbose: 'debug',
    silly: 'debug',
    http: 'info'
} as const;

export function getFullIndexName(prefix: string, createdAt: string) {
    return `${prefix}.${new Date(createdAt).toISOString().split('T')[0]}`;
}

export function createCursor({ sort }: estypes.SearchHit): string {
    return Buffer.from(JSON.stringify(sort)).toString('base64');
}

export function parseCursor(str: string): any[] {
    return JSON.parse(Buffer.from(str, 'base64').toString('utf8'));
}

export const operationTypeToMessage: Record<ConcatOperationList, string> = {
    'action:run': 'Action',
    'admin:impersonation': 'Admin logged into another account',
    'auth:create_connection': 'Connection created',
    'auth:post_connection': 'post connection execution',
    'auth:refresh_token': 'Token refreshed',
    'auth:connection_test': 'Connection test',
    'deploy:custom': 'Deploys',
    'deploy:prebuilt': 'Deploys',
    'proxy:call': 'Proxy',
    'sync:cancel': 'Sync execution canceled',
    'sync:init': 'Sync initialized',
    'sync:pause': 'Sync schedule paused',
    'sync:request_run_full': 'Full execution triggered',
    'sync:request_run': 'Incremental execution triggered',
    'sync:run': 'Sync executed',
    'sync:unpause': 'Sync schedule resumed',
    'webhook:incoming': 'External webhook executed',
    'webhook:forward': 'External webhook forwarded',
    'webhook:sync': 'Sync completion webhooks',
    'webhook:connection_create': 'Connection creation webhooks',
    'webhook:connection_refresh': 'Token refresh webhooks',
    'events:post_connection_creation': 'Event-based executions',
    'events:pre_connection_deletion': 'Event-based executions'
};

/**
 * Send buffered logs to elasticsearch
 * Ultimately it would be better to have LogContextBuffer (or an option in LogContext)
 */
export async function flushLogsBuffer(logs: MessageRowInsert[], logCtx: LogContext) {
    await Promise.all(
        logs.map(async (log) => {
            await logCtx.log(log);
        })
    );
}
