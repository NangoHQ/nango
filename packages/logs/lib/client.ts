import type { MessageRow, MessageRowInsert, MessageMeta, OperationRow, MessageHTTPResponse, MessageHTTPRequest } from '@nangohq/types';
import { setRunning, createMessage, setFailed, setCancelled, setTimeouted, setSuccess, updateOperation } from './models/messages.js';
import { getFormattedMessage } from './models/helpers.js';
import { metrics, report } from '@nangohq/utils';
import { isCli, logger, logLevelToLogger } from './utils.js';
import { envs } from './env.js';
import { OtlpSpan } from './otlp/otlpSpan.js';
import { errorToDocument } from './formatters.js';

interface Options {
    dryRun?: boolean;
    logToConsole?: boolean;
}

/**
 * Context without operation (stateless)
 */
export class LogContextStateless {
    id: OperationRow['id'];
    dryRun: boolean;
    logToConsole: boolean;

    constructor(data: { parentId: OperationRow['id'] }, options: Options = { dryRun: false, logToConsole: true }) {
        this.id = data.parentId;
        this.dryRun = isCli || !envs.NANGO_LOGS_ENABLED ? true : options.dryRun || false;
        this.logToConsole = options.logToConsole ?? true;
    }

    async log(data: MessageRowInsert): Promise<boolean> {
        if (data.error && data.error.constructor.name !== 'Object') {
            data.error = errorToDocument(data.error);
        }

        if (this.logToConsole) {
            const obj: Record<string, any> = {};
            if (data.error) obj['error'] = data.error;
            if (data.meta) obj['meta'] = data.meta;
            logger[logLevelToLogger[data.level!]](`${this.dryRun ? '[dry] ' : ''}log: ${data.message}`, Object.keys(obj).length > 0 ? obj : undefined);
        }
        if (this.dryRun) {
            return true;
        }

        const start = Date.now();
        try {
            await createMessage(getFormattedMessage({ ...data, parentId: this.id }));
            return true;
        } catch (err) {
            report(new Error('failed_to_insert_in_es', { cause: err }));
            return false;
        } finally {
            metrics.duration(metrics.Types.LOGS_LOG, Date.now() - start);
        }
    }

    async debug(message: string, meta?: MessageMeta): Promise<boolean> {
        return await this.log({ type: 'log', level: 'debug', message, meta, source: 'internal', createdAt: new Date().toISOString() });
    }

    async info(message: string, meta?: MessageMeta): Promise<boolean> {
        return await this.log({ type: 'log', level: 'info', message, meta, source: 'internal', createdAt: new Date().toISOString() });
    }

    async warn(message: string, meta?: MessageMeta): Promise<boolean> {
        return await this.log({ type: 'log', level: 'warn', message, meta, source: 'internal', createdAt: new Date().toISOString() });
    }

    async error(message: string, meta: (MessageMeta & { error?: unknown; err?: never; e?: never }) | null = null): Promise<boolean> {
        const { error, ...rest } = meta || {};
        return await this.log({
            type: 'log',
            level: 'error',
            message,
            error: errorToDocument(error),
            meta: Object.keys(rest).length > 0 ? rest : undefined,
            source: 'internal',
            createdAt: new Date().toISOString()
        });
    }

    async http(
        message: string,
        {
            error,
            ...data
        }: {
            request: MessageHTTPRequest | undefined;
            response: MessageHTTPResponse | undefined;
            error?: unknown;
            meta?: MessageRow['meta'];
            level?: MessageRow['level'];
        }
    ): Promise<boolean> {
        const level: MessageRow['level'] = data.level ?? (data.response && data.response.code >= 400 ? 'error' : 'info');
        return await this.log({
            type: 'http',
            level,
            message,
            ...data,
            error: errorToDocument(error),
            source: 'internal',
            createdAt: new Date().toISOString()
        });
    }

    /**
     * @deprecated Only there for retro compat
     */
    async trace(message: string, meta?: MessageMeta): Promise<boolean> {
        return await this.log({ type: 'log', level: 'debug', message, meta, source: 'internal', createdAt: new Date().toISOString() });
    }
}

/**
 * Context with operation (can modify state)
 */
export class LogContext extends LogContextStateless {
    operation: OperationRow;
    span: OtlpSpan;

    constructor(data: { parentId: string; operation: OperationRow }, options: Options = { dryRun: false, logToConsole: true }) {
        super(data, options);
        this.operation = data.operation;
        this.span = new OtlpSpan(data.operation);
    }

    /**
     * Add more data to the parentId
     */
    async enrichOperation(data: Partial<OperationRow>): Promise<void> {
        this.span.enrich(data);
        await this.logOrExec(
            `enrich(${JSON.stringify(data)})`,
            async () => await updateOperation({ id: this.id, data: { ...data, createdAt: this.operation.createdAt } })
        );
    }

    /**
     * ------ State
     */
    async start(): Promise<void> {
        await this.logOrExec('start', async () => await setRunning(this.operation));
    }

    async failed(): Promise<void> {
        await this.logOrExec('failed', async () => {
            await setFailed(this.operation);
            this.span.end('failed');
        });
    }

    async success(): Promise<void> {
        await this.logOrExec('success', async () => {
            await setSuccess(this.operation);
            this.span.end('success');
        });
    }

    async cancel(): Promise<void> {
        await this.logOrExec('cancel', async () => {
            await setCancelled(this.operation);
            this.span.end('cancelled');
        });
    }

    async timeout(): Promise<void> {
        await this.logOrExec('timeout', async () => {
            await setTimeouted(this.operation);
            this.span.end('timeout');
        });
    }

    private async logOrExec(log: string, callback: () => Promise<void>) {
        if (this.logToConsole) {
            logger.info(`${this.dryRun ? '[dry] ' : ''}${log}(${this.id})`);
        }
        if (this.dryRun) {
            return;
        }

        try {
            await callback();
        } catch (err) {
            report(new Error(`failed_to_set_${log}`, { cause: err }));
        }
    }
}
