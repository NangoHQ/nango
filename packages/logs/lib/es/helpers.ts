import { isTest } from '@nangohq/utils';
import { envs } from '../env.js';
import { logger } from '../utils.js';
import { client } from './client.js';
import { getDailyIndexPipeline, indexMessages, policyRetention } from './schema.js';
import { createMessage } from '../models/messages.js';
import { getFormattedMessage } from '../models/helpers.js';

export async function start() {
    if (!envs.NANGO_LOGS_ENABLED) {
        logger.warning('Elasticsearch is disabled, skipping');
        return;
    }

    logger.info('ðŸ”„ Elasticsearch service starting...');

    await migrateMapping();

    logger.info('âœ… Elasticsearch');
}

export async function migrateMapping() {
    try {
        const index = indexMessages;

        logger.info(`Migrating index "${index.index}"...`);

        // -- Policy
        logger.info(`  Updating policy`);
        await client.ilm.putLifecycle(policyRetention());

        // -- Index
        const existsTemplate = await client.indices.existsIndexTemplate({ name: `${index.index}-template` });
        logger.info(`  ${existsTemplate ? 'updating' : 'creating'} index template "${index.index}"...`);

        await client.indices.putIndexTemplate({
            name: `${index.index}-template`,
            index_patterns: `${index.index}.*`,
            template: {
                settings: index.settings!,
                mappings: index.mappings!,
                aliases: { [index.index]: {} }
            }
        });

        // -- Pipeline
        // Pipeline will automatically create an index based on a field
        // In our case we create a daily index based on "createdAt"
        logger.info(`  Updating pipeline`);
        await client.ingest.putPipeline(getDailyIndexPipeline(index.index));

        const existsAlias = await client.indices.exists({ index: index.index });
        if (!existsAlias) {
            // insert a dummy record to create first index
            logger.info(`  Inserting dummy record`);
            await createMessage(getFormattedMessage({ parentId: '-1' }));
        }
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_init_elasticsearch');
    }
}

export async function deleteIndex({ prefix }: { prefix: string }) {
    if (!isTest) {
        throw new Error('Trying to delete stuff in prod');
    }

    try {
        const indices = await client.cat.indices({ format: 'json' });
        await Promise.all(
            indices.map(async (index) => {
                if (!index.index?.startsWith(prefix)) {
                    return;
                }

                await client.indices.delete({ index: index.index, ignore_unavailable: true });
            })
        );
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_deleteIndex');
    }
}
