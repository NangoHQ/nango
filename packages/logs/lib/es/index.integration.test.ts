import { beforeAll, describe, expect, it } from 'vitest';
import { deleteIndex, migrateMapping } from './helpers.js';
import { client } from './client.js';
import { indexMessages } from './schema.js';
import { nanoid } from '@nangohq/utils';
import { createOperation, getOperation, updateOperation } from '../models/messages.js';
import { getFormattedOperation } from '../models/helpers.js';

// This file is sequential
describe('mapping', () => {
    const today = new Date().toISOString().split('T')[0];
    let fullIndexName: string;
    beforeAll(async () => {
        indexMessages.index = `index-messages-${nanoid()}`.toLocaleLowerCase();
        fullIndexName = `${indexMessages.index}.${today}`;

        // Delete before otherwise it's hard to debug
        await deleteIndex({ prefix: 'index-messages' });
    });

    it('should not have an index before migration', async () => {
        await expect(client.indices.getMapping({ index: fullIndexName })).rejects.toThrow();
    });

    it('should migrate', async () => {
        await migrateMapping();
    });

    it('should have create index and alias', async () => {
        await client.indices.getMapping({ index: indexMessages.index });

        await client.indices.getMapping({ index: fullIndexName });
    });

    it('should create one index automatically on log', async () => {
        const today = new Date();
        // Log to automatically create an index
        const id = nanoid();
        await createOperation(getFormattedOperation({ id, operation: { type: 'action', action: 'run' }, createdAt: today.toISOString() }));
        await updateOperation({ id, data: { state: 'failed', createdAt: today.toISOString() } });

        // Should have created a today index
        const mapping = await client.indices.getMapping({ index: fullIndexName });
        expect(mapping[fullIndexName]).toMatchSnapshot();

        const settings = await client.indices.getSettings({ index: fullIndexName });
        expect(settings[fullIndexName]?.settings?.index?.analysis).toMatchSnapshot();
        expect(settings[fullIndexName]?.settings?.index?.sort).toMatchSnapshot();
        expect(settings[fullIndexName]?.settings?.index?.lifecycle).toMatchSnapshot();
    });

    it('should create yesterday index automatically', async () => {
        const yesterday = new Date();
        yesterday.setDate(yesterday.getDate() - 1);
        const yesterdayIndexName = `${indexMessages.index}.${yesterday.toISOString().split('T')[0]}`;

        // Log to automatically create an index
        const id = nanoid();
        await createOperation(getFormattedOperation({ id, operation: { type: 'action', action: 'run' }, createdAt: yesterday.toISOString() }));
        await updateOperation({ id, data: { state: 'failed', createdAt: yesterday.toISOString() } });

        // Should have created a yesterday index
        await client.indices.getMapping({ index: yesterdayIndexName });
        const doc = await getOperation({ id });
        expect(doc.state).toBe('failed');
    });
});
