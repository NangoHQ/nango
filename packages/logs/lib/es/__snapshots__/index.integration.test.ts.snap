// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mapping > should create one index automatically on log 1`] = `
{
  "mappings": {
    "dynamic": "false",
    "properties": {
      "accountId": {
        "type": "keyword",
      },
      "accountName": {
        "type": "keyword",
      },
      "connectionId": {
        "type": "keyword",
      },
      "connectionName": {
        "analyzer": "standard",
        "fields": {
          "keyword": {
            "type": "keyword",
          },
        },
        "type": "text",
      },
      "context": {
        "type": "keyword",
      },
      "createdAt": {
        "type": "date",
      },
      "durationMs": {
        "type": "integer",
      },
      "endUserId": {
        "type": "keyword",
      },
      "endUserName": {
        "type": "keyword",
      },
      "endedAt": {
        "type": "date",
      },
      "environmentId": {
        "type": "keyword",
      },
      "environmentName": {
        "type": "keyword",
      },
      "error": {
        "enabled": false,
        "type": "object",
      },
      "expiresAt": {
        "type": "date",
      },
      "id": {
        "type": "keyword",
      },
      "integrationId": {
        "type": "keyword",
      },
      "integrationName": {
        "analyzer": "standard",
        "fields": {
          "keyword": {
            "type": "keyword",
          },
        },
        "type": "text",
      },
      "jobId": {
        "type": "keyword",
      },
      "level": {
        "type": "keyword",
      },
      "message": {
        "analyzer": "standard",
        "type": "text",
      },
      "meta": {
        "enabled": false,
        "type": "object",
      },
      "operation": {
        "properties": {
          "action": {
            "type": "keyword",
          },
          "type": {
            "type": "keyword",
          },
        },
      },
      "parentId": {
        "type": "keyword",
      },
      "providerName": {
        "type": "keyword",
      },
      "request": {
        "properties": {
          "headers": {
            "enabled": false,
            "type": "object",
          },
          "method": {
            "type": "keyword",
          },
          "url": {
            "type": "keyword",
          },
        },
      },
      "response": {
        "properties": {
          "code": {
            "type": "integer",
          },
          "headers": {
            "enabled": false,
            "type": "object",
          },
        },
      },
      "retry": {
        "properties": {
          "attempt": {
            "type": "integer",
          },
          "max": {
            "type": "integer",
          },
          "waited": {
            "type": "integer",
          },
        },
      },
      "source": {
        "type": "keyword",
      },
      "startedAt": {
        "type": "date",
      },
      "state": {
        "type": "keyword",
      },
      "syncConfigId": {
        "type": "keyword",
      },
      "syncConfigName": {
        "analyzer": "standard",
        "fields": {
          "keyword": {
            "type": "keyword",
          },
        },
        "type": "text",
      },
      "type": {
        "type": "keyword",
      },
      "updatedAt": {
        "type": "date",
      },
      "userId": {
        "type": "keyword",
      },
    },
  },
}
`;

exports[`mapping > should create one index automatically on log 2`] = `
{
  "analyzer": {
    "default": {
      "type": "standard",
    },
    "default_search": {
      "type": "standard",
    },
  },
}
`;

exports[`mapping > should create one index automatically on log 3`] = `
{
  "field": [
    "createdAt",
    "id",
  ],
  "order": [
    "desc",
    "desc",
  ],
}
`;

exports[`mapping > should create one index automatically on log 4`] = `
{
  "name": "policy_retention",
}
`;
